### ISSUE TITLE
mypy is confused by return values from generators

### ISSUE THREAD

**Issue by gvanrossum** [JukkaL/mypy#695]
This is fine:
```
from typing import Iterator

def foo() -> Iterator[int]:
    yield 42

x = next(foo()) + 1000
print(x)
```
But now try this:
```
from typing import Iterator

def foo() -> Iterator[int]:
    yield 42
    return 'abc'
```
This complains about the `return` statement:
```
g.py, line 5: Incompatible return value type: expected typing.Iterator[builtins.int], got builtins.str
```
It's matching the generator return value (which is only accessible by a caller that uses `yield from`) to the declared return type (which is just the type of what you get by calling `foo()`.

**Comment by spkersten** [comment]
I've been fixing this (see spkersten/generator). Type checking return already works in that branch, only `yield from` and some special casing for coroutines need a bit more work.

**Comment by gvanrossum** [comment]
@spkersten, do you still have that branch? If not I'm probably going to look into this -- there also seems overlap with #642 and #1011.

**Comment by gvanrossum** [comment]
FWIW, I think that the example using `return 'abc'` should be rejected, but the error message should indicate that a generator declared to return an Iterator should not return a value in a return statement. However, if the function return is declared as `-> Generator[X, Y, Z]` then the return expression should be type-checked against Z, and the value returned by a yield expression should be type-checked against Y. IOW, ue these mnemonics:
```
def foo() -> Generator[X, Y, Z]:
    yield x
    y = yield
    return z
```
We could treat Iterator[X] as Generator[X, Y, Z] except I still think that even `return None` should be disallowed -- only `return` should be allowed if the function declaration doesn't use Generator.

**Comment by gvanrossum** [comment]
@sakana will address this next. (@Jukka: can you add @sakana as a committer? It's David. :-)

**Comment by ddfisher** [comment]
I'm @ddfisher now!

**Comment by JukkaL** [comment]
Added @ddfisher.
